File: src/main/scala/kim/dencka/scala/medium/ContainerWithMostWater.scala
package kim.dencka.scala.medium

object ContainerWithMostWater extends App {

  def maxArea(height: Array[Int]): Int = {
    @annotation.tailrec
    def run(i: Int, j: Int, max: Int): Int =
      if (i >= j) {
        max
      }
      else {
        val area = (j - i) * math.min(height(i), height(j))
        val newMax = if (area > max) area else max
        val (newI, newJ) = if (height(i) < height(j)) (i + 1, j) else (i, j - 1)
        run(newI, newJ, newMax)
      }

    run(0, height.length - 1, 0)
  }

  println(maxArea(Array(1, 8, 6, 2, 5, 4, 8, 3, 7)))
  println(maxArea(Array(1, 8, 6, 2, 5, 4, 8, 3)))
  println(maxArea(Array(2, 3, 10, 5, 7, 8, 9)))
}

File: src/main/scala/kim/dencka/java/medium/ContainerWithMostWater.java
package kim.dencka.java.medium;

public class ContainerWithMostWater {
    public static void main(String[] args) {
        int[] height = {1, 8, 6, 2, 5, 4, 8, 3, 7};
        System.out.println(maxArea(height)); // 49

        int[] height2 = {1, 3, 2, 5, 25, 24, 5};
        System.out.println(maxArea(height2)); // 24
    }

    public static int maxArea(int[] height) {
        int l = 0, r = height.length - 1;
        int result = Integer.MIN_VALUE;

        while (l < r) {
            int minHeight = Math.min(height[l], height[r]);
            result = Math.max(result, minHeight * (r - l));
            while (l < r && height[l] <= minHeight) l++;
            while (l < r && height[r] <= minHeight) r--;
        }

        return result;
    }
}
